### ISSUE TITLE
MSDN Channel 9 don't return any shows

### ISSUE THREAD

**Issue by matsarv** [retrospect-addon/plugin.video.retrospect#1574]
### Describe the bug

MSDN Channel 9 don't return any shows and the logfile says "process_folder_list returned 0 items"

## Expected Behavior

Display a list of shows

## Actual Behavior

No shows are listed

### To Reproduce

1.  Start MSDN Channel 9

### Debuglog

The **debug** retrospect.log can be found here: https://paste.kodi.tv/oyefojeguf.kodi

### Your Environment
Used Operating system:

 - [X] Android
 - [ ] iOS
 - [ ] Linux
 - [ ] OSX
 - [X] Raspberry-Pi
 - [X] Windows
 - [ ] Windows UWP

Used versions:
- Kodi version: 19.3
- Retrospect version: 5.4.9

**Comment by basrieter** [comment]
Already noticed that. They moved MSDN to Microsoft Learn. Still in doubt whether to rewrite that or to just remove it.

**Comment by basrieter** [comment]
Please try https://github.com/retrospect-addon/plugin.video.retrospect/archive/refs/heads/bug-1574.zip and install from zip in Kodi Matrix.

**Comment by matsarv** [comment]
It works OK on an Raspberry Pi 3B (on a Libreelec with Nexus: LibreELEC-RPi2.arm-11.0-nightly-20211126-7ef3241 - no release of Kodi 19.3) and on a Tablet with Anroid.

When it comes to Kodi on Windows it's not wokring OK.
What is happening:

Starting a video.
Video plays OK to time 00:16 and time stops but video continues for 10 sec.
The video stops and the time jumps forward to 00:26
A round circle with 100 is shown and Kodi frezzes.
After a minute or two the video player close down and the item list is shown again.

Tested on Windows 10 using both Kodi 19.3 and Kodi 20-nightly version with above problem.

Retrospect debug log: https://paste.kodi.tv/ereborejil.kodi

Note! I usually use Libreelec and a big screen (TV). On a PC i use the webbbrowser.

---------------------------------------------------

Two requests:

----------------------------------------------------
If possible, increase the max_items to all.
See below examples.

Python for Beginners 
Episode 1-44
Only shows episodes 15-44 (30 episodes)

More Python for Beginners 
Episode 1-20
OK

Even More Python for Beginners 
Episode 1-31
Only shows episodes 2-31 (30 episodes)

----------------------------------------------------

Show episodes sorted from 1 to last.
Did a test with below code:

Add 
`import re`

Add the following in:

`    def create_video_item(self, result_set):`

Replace Line 157:
`        title = result_set["title"]`

With:
```
        title = result_set["title"]
        if "[" in title:
            reverse = title[::-1]
            start = len(title)-re.search(r'\[', reverse).start()
            end = re.search(r'\]', title).start()
            episode = title[start:end].replace(" of ","x")
            title = "{}. {}".format(episode,title)
```

NOTE! I found one folder that I got an error, the "More Ptyhon for Beginners"


The result with above code:
![Image 2](https://user-images.githubusercontent.com/46671821/143685800-35bb0cd4-29b8-4a7d-ae72-e3b40b1838c8.png)

**Comment by basrieter** [comment]
Ok, so what I will do is:

- [ ] Add a "more items"-item so you can get to the older episodes.
- [ ] Try to set the season/episode information if present so you can sort.

The playback issue is not a Retrospect thing. If it starts, the Retrospect is done and it is a Kodi thing.

**Comment by basrieter** [comment]
Part 2 is done. 

![image](https://user-images.githubusercontent.com/1310443/143707642-60853b00-a992-4b4a-8580-cd6c801c14ee.png)

Will try to do the "More items" later.

**Comment by basrieter** [comment]
if you download and install https://github.com/retrospect-addon/plugin.video.retrospect/archive/refs/heads/bug-1574.zip again the things should be fixed.

**Comment by matsarv** [comment]
Now all the episodes are listed.

I do not see the numbering of the episodes.
If I know that there are episodes in the list then I can manually change the Sort by to Episode. My guess is that this should be handled by Framework.

---------------------
First proposal
Use sort method for Episode to be used in KODI.

---------------------


Now it looks like this:
![1a](https://user-images.githubusercontent.com/46671821/143733033-acbbe3f4-c7a0-4b67-a38f-8766f0952898.png)


Made the following changes to the Framework:

Sort method for Episode is not set to be used in KODI.

mediaitem.py
------------

In function:
    def add_stream(self, url, bitrate=0, subtitle=None):

Add after Line 197:
        return MediaItem.LabelTrackNumber in self.__infoLabels

Add:
```
    def has_episode(self):
        """ Does this MediaItem have a Episode InfoLabel

        :return: if the episode was set.
        :rtype: bool
        """

        return MediaItem.LabelEpisode in self.__infoLabels
```

mediaitem.py
------------

In class:
class MediaItem:

Add after Line 35:
    LabelTrackNumber = "TrackNumber"

Add:
```
     LabelEpisode = "Episode"
```


folderaction.py
---------------

In function:
    def __add_sort_method_to_handle(self, handle, items=None):

After Line 272:
                sort_methods.insert(0, xbmcplugin.SORT_METHOD_TRACKNUM)

Add:
```
            has_episodes = any([i for i in items if i.has_episode()])
            if has_episodes:
                sort_methods.insert(0, xbmcplugin.SORT_METHOD_EPISODE)
```


The numbers of the episodes are now displayed and sorted.
This happens in all folders that have episodes.

![1b](https://user-images.githubusercontent.com/46671821/143733219-d0e4af49-6ac3-459d-b9bf-ec70877c228d.png)


This is just a suggestion so that this can (should?) be used where you have seasons and episodes in a folder.


---------------------
Second proposal 
Add correct information and layout to the videoplayer OSD.

---------------------

The following is now displayed on the KODI video player OSD:

For Season/Episode
![2a](https://user-images.githubusercontent.com/46671821/143733736-a8973dc8-44dc-4abe-9787-26e09eddf93f.png)

For single shows
![3a](https://user-images.githubusercontent.com/46671821/143733762-7d41643d-cfdc-48ea-9167-4c70eeefd171.png)

Made the following changes to the channel

Replaced all mediatype.EPISODE to mediatype.VIDEO and when the content is season/episode I added mediatype.EPISODE and the set_info_label "tvshowtitle" to the TV Show title.


In function:
    def create_video_item(self, result_set):

Replace from Line 194:
```
        item = MediaItem(title, url, media_type=mediatype.EPISODE)
        item.description = description or ""
        if season and episode:
            item.set_season_info(season, episode)
```

With:
```
        item = MediaItem(title, url, media_type=mediatype.VIDEO)
        item.description = description or ""
        if season and episode:
            item.media_type = mediatype.EPISODE
            tvshow_title =  self.parentItem.name
            item.set_info_label("tvshowtitle", tvshow_title)
            item.set_season_info(season, episode)
```

For Season/Episode
![2b](https://user-images.githubusercontent.com/46671821/143733913-80c44108-7b81-4d75-993f-a3464c618131.png)

For single shows
![3b](https://user-images.githubusercontent.com/46671821/143733918-67fe6c03-3cd9-452c-a0de-5c958349e447.png)


My suggestion is to use the following for season/episode in all channels.

Listing:
![4a](https://user-images.githubusercontent.com/46671821/143734153-c9aa90af-727e-4cde-94e4-5301897b166d.png)

OSD:
![4b](https://user-images.githubusercontent.com/46671821/143734155-9f328a86-cc9e-4939-9871-b1f1dd9411cb.png)


--------------------------------
Note!
This is just one way and a suggestion from my side.
The above behavior is present in many channels and changes will have an impact. 
Tested in SVT Play, TV4 Play and Viafree.

**Comment by basrieter** [comment]
Great! The sorting makes sense. Will add that.

The TV Show title is already handled in #1577.

**Comment by matsarv** [comment]
Short info!

I have checked the brach bug-1577 and installed and tested it.
This has a different approach to add TVShowTitle

This work as long as you don't have series with deeper structure.

For example, tested on SVT Play:
1. Select SVT Play
2. Select Categories
3. Select Serier
4. Select Säsong 1

Playing the first episode will resuld in: 

With fix bug-1577
![10a](https://user-images.githubusercontent.com/46671821/143764122-16ebba9a-6d96-4052-9ec9-996c37a799f1.png)

My test with changes in chn_svtplay.py
![10b](https://user-images.githubusercontent.com/46671821/143764128-9efd0733-2c5c-4335-8b2d-82ea0781d045.png)

**Comment by matsarv** [comment]
As for the Short info! above I have tested a bit more and the fix you did (bug-1577) to download title automatically works very well.
For now, I think it's OK that it will be "Season 1" as the title on some channel series. This can be easily fixed at a later time by adding TVShowTitle to the channel file and then this title will override the automatic title. Greate work.

**Comment by basrieter** [comment]
Could you redownload https://github.com/retrospect-addon/plugin.video.retrospect/archive/refs/heads/bug-1574.zip and try again?

**Comment by basrieter** [comment]
> I have tested the above and everyting seems to be working ok. After the Framework changes, some channels need some minor update and I will write new issues after merged.

Ok good to hear. Please submit actual PR's for each individual channel. That makes reviewing/merging much more efficient.

**Comment by basrieter** [comment]
Closed with ebdb3debd74b742364432c1431389cdc02d92805
